src/com/selftest/queue/AbstractQueue.java

```diff
@@ -0,0 +1,56 @@
+
+package com.selftest.queue;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public abstract class AbstractQueue<T> implements Queue<T> {
+
+	protected T[] queue;
+	protected int putLoc, getLoc;
+
+	@SafeVarargs
+	public AbstractQueue( int size, T ... values ) {
+		queue = Arrays.copyOf( values, size );
+		putLoc = getLoc = 0;
+	}
+
+	@Override
+	public synchronized Iterator<T> iterator() {
+		return new Iterator<T>() {
+
+			int position = 0;
+
+			@Override
+			public boolean hasNext() {
+				return position < queue.length;
+			}
+
+			@Override
+			public T next() {
+				return queue[position++];
+			}

```

Comment:
As it happens, Java mandates reference writes be atomic [\[1\]](http://docs.oracle.com/javase/specs/jls/se5.0/html/memory.html#17.7), so this is a safe iterator :smile:.